lib/Parse/ParseExpr.cpp

```diff
@@ -2399,6 +2399,39 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
     }
   }
 
+  if (!params)
+    return invalid;
+
+  // If this was a closure declaration (maybe even trailing)
+  // it's imperative to detect any parameter list destructuring
+  // early on, see SE-0110 for more details.
+  for (unsigned i = 0, e = params->size(); i != e; ++i) {
+    auto *param = params->get(i);
+    if (!param->isInvalid())
+      continue;
+
+    auto paramRange = CharSourceRange(
+        SourceMgr, params->getStartLoc(),
+        Lexer::getLocForEndOfToken(SourceMgr, params->getEndLoc()));
+
+    SmallString<64> fixIt;
+    llvm::raw_svector_ostream OS(fixIt);
+
+    if (Tok.isAtStartOfLine())
+      OS << '\n';
+
+    OS << "let " << SourceMgr.extractText(paramRange) << " = arg"
+       << (Tok.isAtStartOfLine() ? '\n' : ';');
+
+    diagnose(params->getStartLoc(), diag::anon_closure_arg_tuple_destructuring)
+        .fixItReplace(params->getSourceRange(), "(arg)")
+        .fixItInsertAfter(Lexer::getLocForEndOfToken(SourceMgr, inLoc),
```

Comment:
Instead of "after the `in`", this should be "before `Tok`" (the first token of the body, right?). I'm not sure how to preserve the indentation for the actual first line, though.